### ISSUE TITLE
prerelease_script: Problem with looking up valid OS codes

### ISSUE THREAD

**Issue by mani-monaj** [ros-infrastructure/ros_buildfarm#90]
Apparantly there is a problem in looking up valid OS codes when creating pre-release jobs for distros other than local build machine's distro. My local build machine runs 64 bit _Ubuntu Trusty_ and has been setup to run prerelease jobs based on [the documentation on ROS wiki](http://wiki.ros.org/regression_tests#How_do_I_setup_my_system_to_run_a_prerelease.3F).

The prerelease command works fine when target distro is set to _Trusty_. However, when the target distro is set to _Vivid_ or _Utopic_ the generated command fails with the following error:

```
$ generate_prerelease_script.py \
>   https://raw.githubusercontent.com/ros-infrastructure/ros_buildfarm_config/master/index.yaml \
>   jade default ubuntu utopic amd64 \
>   ardrone_autonomy \
>   --level 1 \
>   --output-dir .
Fetching buildfarm configuration...
Fetching rosdistro cache...
Evaluating job templates...
Traceback (most recent call last):
  File "/usr/bin/generate_prerelease_script.py", line 263, in <module>
    sys.exit(main())
  File "/usr/bin/generate_prerelease_script.py", line 157, in main
    source_repository=source_repository)
  File "/usr/lib/python2.7/dist-packages/ros_buildfarm/devel_job.py", line 237, in configure_devel_job
    ', '.join(sorted(build_file.targets[os_name].keys())))
ros_buildfarm.common.JobValidationError
```

Is there any way to setup a local and [minimal development environment to debug this issue? If so, I can try to track down the bug and fix it.

**Comment by tfoote** [comment]
This is a standard python package. You can clone it and prepend it's module directory to your PYTHONPATH and the scripts directory to your PATH. 

You can also use a full virtualenv for development and testing. It's more than you need but provides clear separation.

**Comment by mani-monaj** [comment]
@tfoote Thanks. I'll work on this then.

**Comment by mani-monaj** [comment]
Target OS keys for pre-release tests are fetched from [`ros_buildfarm_config/jade/source-build.yaml`](https://github.com/ros-infrastructure/ros_buildfarm_config/blob/3dfd9fd5f59f348554e517301decd036cfc7ff00/jade/source-build.yaml).  Currently 64 bit _Ubuntu Trusty_ is the only target in that configuration files. That's why the lookup fails when target is set to anything other than _Trusty_.

Is this a design decision to limit source builds of _Jade_ to Ubuntu Trusty? If no, I think this ticket needs to be moved to [`ros-infrastructure/ros_buildfarm_config`](https://github.com/ros-infrastructure/ros_buildfarm_config) issues.

**Comment by tfoote** [comment]
That config defines what target architectures to run the devel jobs for continuous integration. We should probably be using the binary target architectures for the supported set of prereleases across all binary release files (aka get both the x86 and the armhf builds). https://github.com/ros-infrastructure/ros_buildfarm_config/blob/3dfd9fd5f59f348554e517301decd036cfc7ff00/jade/release-armhf-build.yaml and https://github.com/ros-infrastructure/ros_buildfarm_config/blob/3dfd9fd5f59f348554e517301decd036cfc7ff00/jade/release-build.yaml

**Comment by mani-monaj** [comment]
Thanks for the clarification. These are two approaches that come to my mind to fix this issue:

1. Change `[jade|indigo]/source-build.yaml` config file to include other Ubuntu distro and architecures?
2. Add a section to `index.yaml` config file for pre-release tests. We can re-use the `release-build` config or create a new config file for pre-releases only.

```yaml
distributions:
    [...]
  jade:
    [...]
    source_builds:
      default: jade/source-build.yaml
      pre-release: jade/release-build.yaml or jade/prerelease-build.yaml
```

For this approach to work, the web app that generates the pre-release command also needs to be patched.

**Comment by mani-monaj** [comment]
I quickly tried the second approach by forking the config repo and modifying `index.yaml` to re-use `release-build` config. It fails in `source_build_file.py` with this error: `AssertionError: Expected file type is 'source-build', not 'release-build'` (The types are indeed different)

**Comment by mani-monaj** [comment]
Another quick test: I created a new config file for pre-release builds, then changed `index.yaml` to use that config file for source builds. This new file is a copy of `source-builds.yaml` with its `targets` section set to:

```yaml
targets:
  ubuntu:
    trusty:
      amd64:
      i386:
    utopic:
      amd64:
      i386:
    vivid:
      amd64:
      i386:
```

The pre-release script works fine with the new config file for _Trusty_ and _Utopic_. It fails on _Vivid_, due to some Python errors that seem to be unrelated to the current issue.

**Comment by tfoote** [comment]
As this is a prerelease I think that reusing the release_builds config element makes sense. There's not much point in doing pre-release test on something you're not going to be able to release. The error you hit is correct as you listed the release build type files in the source build space. But if we instead make it iterate the release-build configs instead of the source build configs it should work out better and we don't need to define a new filetype or label.

**Comment by mani-monaj** [comment]
Sounds good. I'll make the required changes to iterate through the release-build configs for pre-release tests and submit a PR if everything works well.

**Comment by mani-monaj** [comment]
@tfoote Unless I am missing something, iterating through release-build configs for pre-release jobs is not trivial.

I tried changing instances of `get_source_build_files()`  in `generate_prerelease_script.py` to `get_release_build_files()` so the `build_file` object is generated from release-build configs. Since the resulting object type is different, `configure_devel_job()` fails with the following error:

```
AttributeError: 'ReleaseBuildFile' object has no attribute 'jenkins_commit_job_priority'
```

I also tried changing `generate_prerelease_script.py` to create `release` jobs instead of `devel` jobs. It fails as well since it tries to interact with _Jenkins_.
